sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JSONRelation.scala

```diff
@@ -148,10 +147,7 @@ class DefaultSource extends FileFormat with DataSourceRegister {
         columnNameOfCorruptRecord,
         parsedOptions)
 
-      val appendPartitionColumns = GenerateUnsafeProjection.generate(fullSchema, fullSchema)
-      rows.map { row =>
-        appendPartitionColumns(joinedRow(row, file.partitionValues))
-      }
+      FileFormat.appendPartitionValues(rows, fullSchema, file.partitionValues)
```

Comment:
We will send the reader function to executors, and run it many times for all the files in that RDD partition. However, the projection we used to append partition values can be only initialized once.

Should we extend the reader function to make it have an `initialize` method? e.g. we can create a `trait ReaderFunction` and define the `initialize` and `execute` API.